### PR TITLE
GH-6573 UpliftDRF Variable Importance

### PR DESCRIPTION
**Pull request by maurever**
Issue:  #6573 

Here are mentioned two approaches to calculating variable importance: https://pbiecek.github.io/xai_stories/story-uplift-marketing1.html#difference---approach

However, we can also calculate variable importance based on Uplift Gain, calculated for every split. I decided to use this approach as a baseline. In the future, we can implement more approaches. WDYT?

### CODE REVIEW

**Review by wendycwong — APPROVED** [review]

File: h2o-algos/src/main/java/hex/tree/DTree.java [hunk 1]
```diff
@@ -1652,20 +1662,21 @@ static Split findBestSplitPointUplift(DHistogram hs, int col, double min_rows) {
...
     double prY1CT0 = nCT0Y1hi/nCT0;
     double bestUpliftGain = upliftMetric.node(prY1CT1 , prY1CT0);
+    double upliftGainBefore = bestUpliftGain;
     // if there are any NAs, then try to split them from the non-NAs
     if (wNA>=min_rows) {
       double prCT1All = (nCT1 + numTreatNA + 1)/(nCT0 + numContrNA + nCT1 + numTreatNA + 2);
       double prCT0All = 1-prCT1All;
-      double prY1CT1All = (resphiTreat[0] + respTreatNA) / (nCT1 + numTreatNA);
-      double prY1CT0All = (resphiContr[0] + respContrNA) / (nCT0 + numContrNA);
+      double prY1CT1All = (nCT1Y1hi + respTreatNA) / (nCT1 + numTreatNA);
```
Comment (tomasfryda) [inline h2o-algos/src/main/java/hex/tree/DTree.java:1670]: I know it probably seems obvious for you what the `nCT1Y1hi` means but would you please at least comment near the variable declaration what it is supposed to mean, i.e., how to decode the name to some meaningful description (e.g., imagine I have a variable `nCY1` which can be decoded as  "number of Control data points with Y ==1").

To be clear, I am perfectly ok with using short variable names if it is easy to decode them.

File: h2o-algos/src/test/java/hex/tree/uplift/UpliftDRFTest.java [hunk 2]
```diff
@@ -75,7 +77,7 @@ public void testBasicTrainAndScore() {
             assertArrayEquals(new String[]{"uplift_predict", "p_y1_with_treatment", "p_y1_without_treatment"}, out.names());
             assertEquals(train.numRows(), out.numRows());
 
-            assertNull(model._output._varimp); // not supported yet
+            assertNotNull(model._output._varimp); // not supported yet
```
Comment (tomasfryda) [inline h2o-algos/src/test/java/hex/tree/uplift/UpliftDRFTest.java:80]:
```suggestion
            assertNotNull(model._output._varimp);
```

**Review by tomasfryda — APPROVED** [review]
Looks good. I have just some minor concerns. Thank you @maurever !

One additional question of mine is how useful is this variable importance for the end user. My assumption is that people use variable importance to gain more confidence in the model. For ordinary models variable importance makes sense but when considering causal models wouldn't it make more sense to show variable importance conditioned on treatment?  I guess you probably want the model to have similar importances in both groups so you can be more confident that the model doesn't just use some specific information to find out if there was a treatment or not. I know very little about these kinds of models but if you think there might be a use-case similiar to what I described, maybe it would be useful to test if it's possible to use uplift with permutaion variable importance and document using it with mask `treatment=1` and `treatment=0`. WDYT?

File: h2o-docs/src/product/data-science/upliftdrf.rst [hunk 3]
```diff
@@ -388,6 +389,15 @@ Average Excess Cumulative Uplift (AECU)
 
 The Qini value can be generalized for all AUUC metric types. So AECU for Qini metric is the same as Qini value, but the AECU can be also calculated for Gain and Lift metric type. These values are stored in ``aecu_table``.
 
+
+Variable Importance
+~~~~~~~~~~~~~~~~~~~
+
+The Variable Importance is calculated based on Uplift Gain improvement. After every split the local split improvement is calculated as difference between Uplift Gain before split and Uplift Gain after split. 
+
+The Uplift Gain is calculated based on selected ``uplift_metric``. The higher Uplift Gain the better.
```
Comment (valenad1) [inline h2o-docs/src/product/data-science/upliftdrf.rst:398]: Do we have some external sources that we can share here?

Comment (valenad1) [reply h2o-docs/src/product/data-science/upliftdrf.rst:398]: Or just share the formula we are using in the code.

File: h2o-algos/src/test/java/hex/tree/uplift/UpliftDRFTest.java [hunk 4]
```diff
@@ -471,6 +473,33 @@ public void testEasyPredictMojo() throws Exception {
             Scope.exit();
         }
     }
+
+    @Test
+    public void testVariableImportance(){
```
Comment (valenad1) [inline h2o-algos/src/test/java/hex/tree/uplift/UpliftDRFTest.java:478]: I saw that all the test are testing if variable importance is not null. Can we compare the values with some reference or syntetic dataset to see if the computation is correct and make sense?

File: h2o-docs/src/product/data-science/upliftdrf.rst (hunk 3 above)
Comment (maurever) [reply h2o-docs/src/product/data-science/upliftdrf.rst:398]: The formulas are described in the ``uplift_metric`` section. :)

**Comment by maurever** [comment]
> Looks good. I have just some minor concerns. Thank you @maurever !
> 
> One additional question of mine is how useful is this variable importance for the end user. My assumption is that people use variable importance to gain more confidence in the model. For ordinary models variable importance makes sense but when considering causal models wouldn't it make more sense to show variable importance conditioned on treatment? I guess you probably want the model to have similar importances in both groups so you can be more confident that the model doesn't just use some specific information to find out if there was a treatment or not. I know very little about these kinds of models but if you think there might be a use-case similiar to what I described, maybe it would be useful to test if it's possible to use uplift with permutaion variable importance and document using it with mask `treatment=1` and `treatment=0`. WDYT?

Thanks @tomasfryda for the suggestions. In the uplift gain, we are computing the difference between treatment and control group probabilities, so I think this metric is useful to decide how the model improves based on a split with a feature. But there are definitely plenty of other methods how compute variable importance. I decided to use this approach as a baseline because it is easy to implement and makes sense to me.  I mentioned in the PR description we can implement other approaches later.  WDYT?

File: h2o-algos/src/test/java/hex/tree/uplift/UpliftDRFTest.java (hunk 4 above)
Comment (maurever) [reply h2o-algos/src/test/java/hex/tree/uplift/UpliftDRFTest.java:478]: Thanks for the suggestion! You are right. I can improve the test in the next fix release. WDYT @valenad1?

File: h2o-algos/src/main/java/hex/tree/DTree.java (hunk 1 above)
Comment (maurever) [reply h2o-algos/src/main/java/hex/tree/DTree.java:1670]: Sure, I will do it. Thanks!

File: h2o-docs/src/product/data-science/upliftdrf.rst (hunk 3 above)
Comment (valenad1) [reply h2o-docs/src/product/data-science/upliftdrf.rst]: IMHO we are mixing terminologies. In uplift metric is written:

`Based on the uplift_metric parameter, the distribution divergence is calculated`

IMHO uplift_metric calculates distribution divergence == uplift gain. We store the uplift gain (divergence) for each variable.

Variable importance is sum of those gains.

Am I right? If yes, than I have to look in the code, i would like to be sure just from documentation.

Is it our idea or we can share some paper on it? I remember, the link to the paper in the code saved me one time.

**Review by wendycwong — APPROVED** [review]

**Review by tomasfryda — APPROVED** [review]

**Pull request merged**